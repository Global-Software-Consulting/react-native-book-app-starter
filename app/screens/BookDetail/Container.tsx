import React, { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { NativeEventEmitter, NativeModules, ScrollView, View } from 'react-native';
import { Button, Text } from 'react-native-paper';
import Icon from 'react-native-vector-icons/FontAwesome';
import BookCard from '../../components/BookCard/BookCard';
import Tts from 'react-native-tts';
import { useStyles } from './styles';
import { Props } from './types';
import { useNavigation } from '@react-navigation/core';
import { TouchableOpacity } from 'react-native-gesture-handler';

const Container: React.FC<Props> = (props) => {
    const dummyImages = [
        'https://d1csarkz8obe9u.cloudfront.net/posterpreviews/sad-romance-kindle-book-cover-flyer-template-539b5fdf2dd2f6602c25ce81fbb5d877.jpg?ts=1589326539',
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSO4yGwkz128dpVHBztwERbm6Z6kIXwQ03V0A&usqp=CAU',
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQQB-3-cJ6tpQ6cFUjCach0R1dTiXdhXI0AtA&usqp=CAU',
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRuaVjRvjIZAO6XIsd0eHS22Q949ke7r-KZFg&usqp=CAU',
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRkIVlYNBoYdfq1FGWs_dx_k8Irvjmdu64p7A&usqp=CAU',
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRwor6J4qPx1WWN27GYiZEFP-NAoWSstn-YrQ&usqp=CAU',
    ];

    const generateRandomURL = () => {
        const number = Math.random() * 6;
        const index = Math.floor(number);
        return dummyImages[index];
    };
    //theme handling
    const styles = useStyles();
    const navigation = useNavigation();
    const { books } = props;
    const author = 'Dummy author';
    const genre = 'Dummy Genre';
    const pages = books.numberOfPages ?? '111';
    const launched = books.createdAt ?? '11-11-11';
    const title = books.title ?? 'Dummy Title';
    const { t } = useTranslation();
    const [textShown, setTextShown] = useState(false); //To show ur remaining Text
    const [isSpeaking, setIsSpeaking] = useState(false);

    useEffect(() => {
        const ee = new NativeEventEmitter(NativeModules.TextToSpeech);
        ee.addListener('tts-start', () => {});
        ee.addListener('tts-finish', () => {});
        ee.addListener('tts-cancel', () => {});
        if (isSpeaking === true) {
            Tts.speak(books.shortSummary);
        } else {
            Tts.stop();
        }
    }, [isSpeaking]);

    const toggleNumberOfLines = () => {
        //To toggle the show text or hide it
        setTextShown(!textShown);
    };

    return (
        <View style={styles.main}>
            <ScrollView>
                <BookCard
                    id={books?.id ?? 1}
                    styleSelect="ExtraLarge"
                    bookTitle={title}
                    authorName={author}
                    hideIcon={true}
                    url={generateRandomURL()}
                />
                <View
                    style={{
                        flexDirection: 'row',
                        alignItems: 'center',
                        justifyContent: 'center',
                    }}>
                    <View
                        style={[
                            styles.submitView,
                            { backgroundColor: !isSpeaking ? '#00416A' : 'red' },
                        ]}>
                        <Button
                            onPress={() => {
                                setIsSpeaking(!isSpeaking);
                            }}
                            style={styles.submit}>
                            <Text style={{ color: 'white' }}>
                                {!isSpeaking ? t('Speak') : t('Stop')}
                            </Text>
                        </Button>
                    </View>
                    <View style={[styles.submitView, { backgroundColor: '#00416A' }]}>
                        <Button
                            onPress={() => navigation.navigate('BookReader' as never)}
                            style={styles.submit}>
                            <Text style={{ color: 'white' }}>Read Book</Text>
                        </Button>
                    </View>
                </View>
                <View style={styles.horizontalRuler} />

                <View style={styles.mainView}>
                    <View style={styles.subView}>
                        <Text style={styles.genreText}>{t('Genre')}</Text>
                        <Text style={styles.dynamicGenreText}>{genre}</Text>
                    </View>

                    <View style={styles.launchedSubView}>
                        <Text style={styles.launchedText}> {t('Launched')}</Text>
                        <Text style={styles.dynamicLaunchedText}>{launched.substring(0, 9)}</Text>
                    </View>

                    <View style={styles.pagesSubView}>
                        <Text style={styles.sizeText}>{t('Size')}</Text>
                        <Text style={styles.dynamicSize}>
                            {pages}
                            {t(' pages')}
                        </Text>
                    </View>
                </View>

                <View style={styles.synopseView}>
                    <View style={{ flexDirection: 'row' }}>
                        <Text style={styles.synopseText}>{t('Synopse')}</Text>
                    </View>
                    <Text
                        style={styles.dynamicSynopse}
                        onPress={toggleNumberOfLines}
                        numberOfLines={textShown ? undefined : 2}>
                        {books.shortSummary}
                    </Text>
                </View>
            </ScrollView>
        </View>
    );
};

export default Container;
