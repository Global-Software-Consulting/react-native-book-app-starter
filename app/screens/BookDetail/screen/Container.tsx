import React, {useState} from 'react';
import {useTranslation} from 'react-i18next';
import {ScrollView, StyleSheet, View} from 'react-native';
import {Text} from 'react-native-paper';
import {useStyles} from '../styles';
import BookCard from '../../../components/BookCard/BookCard';
import i18n from '../../../config/Languages';
import {heightPercentageToDP} from 'react-native-responsive-screen';

//importing card component
interface Props {
  books?: [];
  base_url?: string;
}
const initI18n = i18n;

const Container: React.FC<Props> = ({books, base_url}) => {
  //theme handling
  const styles = useStyles();
  var author = books.bookAuthors[0].author.name;
  var genre = 'Dummy Genre';
  var pages = books.numberOfPages;
  var launched = books.createdAt;
  var title = books.title;
  const {t, i18n} = useTranslation();

  const [textShown, setTextShown] = useState(false); //To show ur remaining Text
  const [lengthMore, setLengthMore] = useState(false); //to show the "Read more & Less Line"
  const toggleNumberOfLines = () => {
    //To toggle the show text or hide it
    setTextShown(!textShown);
  };
  console.log('Book detailo', books);
  return (
    <View style={styles.main}>
      <ScrollView>
        <BookCard
          styleSelect="ExtraLarge"
          bookTitle={title}
          authorName={author}
          hideIcon={true}
          url="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRoE4lMLbADvLAxUvZf5ZAGvHUZ3KpBWFTW1g&usqp=CAU"
        />

        <View style={styles.horizontalRuler} />

        <View style={styles.mainView}>
          <View style={styles.subView}>
            <Text style={styles.genreText}>{t('Genre')}</Text>
            <Text style={styles.dynamicGenreText}>{genre}</Text>
          </View>

          <View style={styles.launchedSubView}>
            <Text style={styles.launchedText}> {t('Launched')}</Text>
            <Text style={styles.dynamicLaunchedText}>
              {launched.substring(0, 9)}
            </Text>
          </View>

          <View style={styles.pagesSubView}>
            <Text style={styles.sizeText}>{t('Size')}</Text>
            <Text style={styles.dynamicSize}>
              {pages}
              {t(' pages')}
            </Text>
          </View>
        </View>

        <View style={styles.synopseView}>
          <Text style={styles.synopseText}>{t('Synopse')}</Text>
          <Text
            style={styles.dynamicSynopse}
            onPress={toggleNumberOfLines}
            numberOfLines={textShown ? undefined : 2}>
            {books.shortSummary}
          </Text>
        </View>
      </ScrollView>
    </View>
  );
};

export default Container;
