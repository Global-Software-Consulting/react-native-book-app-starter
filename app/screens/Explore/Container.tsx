import { useNavigation } from '@react-navigation/native';
//importing card component
import BookCard from 'components/BookCard/BookCard';
import images from 'config/images';
import { Books, IAppState } from 'models/reducers/appReducers';
import React from 'react';
import { useTranslation } from 'react-i18next';
import {
    FlatList,
    Image,
    RefreshControl,
    ScrollView,
    TouchableHighlight,
    View,
} from 'react-native';
import { Text, useTheme } from 'react-native-paper';
import { useSelector } from 'react-redux';
import { useStyles } from 'screens/Explore/styles';
import { Props } from './types';
const ExploreComponent: React.FC<Props> = (props) => {
    //theme handling
    const styles = useStyles();
    const { t } = useTranslation();
    const theme = useTheme();
    const navigation = useNavigation();
    const { name, onRefresh } = props;
    const dummyImages = [
        'https://d1csarkz8obe9u.cloudfront.net/posterpreviews/sad-romance-kindle-book-cover-flyer-template-539b5fdf2dd2f6602c25ce81fbb5d877.jpg?ts=1589326539',
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSO4yGwkz128dpVHBztwERbm6Z6kIXwQ03V0A&usqp=CAU',
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQQB-3-cJ6tpQ6cFUjCach0R1dTiXdhXI0AtA&usqp=CAU',
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRuaVjRvjIZAO6XIsd0eHS22Q949ke7r-KZFg&usqp=CAU',
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRkIVlYNBoYdfq1FGWs_dx_k8Irvjmdu64p7A&usqp=CAU',
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRwor6J4qPx1WWN27GYiZEFP-NAoWSstn-YrQ&usqp=CAU',
    ];

    const generateRandomURL = () => {
        const number = Math.random() * 6;
        const index = Math.floor(number);
        return dummyImages[index];
    };
    //getting data from store
    const books = useSelector((state: { appReducer: IAppState }) => state.appReducer.books);
    const isLoading = useSelector(
        (state: { loadingReducer: IAppState }) => state.loadingReducer.isLoading,
    );

    const navigateToDetails = (params: number) => {
        navigation.navigate('BookDetail' as never, params as never);
    };

    const bookList = (item: Books, style: 'General' | 'Custom') => {
        return (
            <TouchableHighlight
                key={item.id}
                underlayColor={theme.colors.highlight}
                onPress={() => {
                    navigateToDetails(item.id);
                }}>
                <BookCard
                    url={generateRandomURL()}
                    styleSelect={style}
                    bookTitle={item?.title}
                    book={item}
                    id={item?.id}
                />
            </TouchableHighlight>
        );
    };

    return (
        <View style={styles.mainViewSetting}>
            <ScrollView
                nestedScrollEnabled={true}
                style={styles.container}
                showsVerticalScrollIndicator={false}
                refreshControl={
                    <RefreshControl
                        refreshing={isLoading}
                        onRefresh={onRefresh}
                        colors={['green', 'yellow']}
                    />
                }>
                <View style={styles.middleView}>
                    <Text style={styles.name}>
                        {t('Hi')} {name}{' '}
                    </Text>
                    <Text style={styles.tagLine}>{t('Lets find something new')}</Text>

                    {books?.length > 0 ? (
                        <View>
                            <View style={styles.horizontalRuler} />
                            <Text style={styles.listCaption}>{t('Trending')}</Text>
                            <FlatList
                                nestedScrollEnabled={true}
                                horizontal
                                data={books?.filter((item: Books) => {
                                    return item.averageRating > 3;
                                })}
                                contentContainerStyle={styles.flatList}
                                renderItem={({ item }) => bookList(item, 'Custom')}
                                showsHorizontalScrollIndicator={false}
                                keyExtractor={(item, index) => 'key' + index}
                            />
                            <View style={styles.horizontalRuler} />
                            <Text style={styles.listCaption}>{t('New Releases')}</Text>
                            <FlatList
                                horizontal
                                data={books?.filter((item: Books) => {
                                    return item.averageRating <= 3 && item.averageRating > 0;
                                })}
                                contentContainerStyle={styles.flatList}
                                renderItem={({ item }) => bookList(item, 'General')}
                                showsHorizontalScrollIndicator={false}
                                keyExtractor={(item, index) => 'key' + index}
                            />
                            <View style={styles.horizontalRuler} />
                            <Text style={styles.listCaption}>{t('Selected for you')}</Text>
                            <FlatList
                                horizontal
                                data={books?.filter((item: Books) => {
                                    return item.averageRating === 0;
                                })}
                                contentContainerStyle={styles.flatListLast}
                                renderItem={({ item }) => bookList(item, 'General')}
                                showsHorizontalScrollIndicator={false}
                                keyExtractor={(item, index) => 'key' + index}
                            />
                        </View>
                    ) : (
                        <View style={styles.favoriteView}>
                            <Image source={images.books.noBookFound} style={styles.imageError} />
                            <Text style={styles.bookmark}>No books found</Text>
                        </View>
                    )}
                </View>
            </ScrollView>
        </View>
    );
};

export default ExploreComponent;
