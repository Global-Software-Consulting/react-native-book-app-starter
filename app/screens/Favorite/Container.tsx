import { useNavigation } from '@react-navigation/native';
//importing card component
import BookCard from 'components/BookCard/BookCard';
import images from 'config/images';
import { FavoriteBook } from 'models/reducers/appReducers';
import { ReducerState } from 'models/reducers/index';
import React, { useState } from 'react';
import { Dimensions, FlatList, Image, ScrollView, TouchableHighlight, View } from 'react-native';
import { Text, useTheme } from 'react-native-paper';
import { useSelector } from 'react-redux';
import { useStyles } from './styles';
import { Props } from './types';

const Container: React.FC<Props> = (props) => {
    const dummyImages = [
        'https://d1csarkz8obe9u.cloudfront.net/posterpreviews/sad-romance-kindle-book-cover-flyer-template-539b5fdf2dd2f6602c25ce81fbb5d877.jpg?ts=1589326539',
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSO4yGwkz128dpVHBztwERbm6Z6kIXwQ03V0A&usqp=CAU',
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQQB-3-cJ6tpQ6cFUjCach0R1dTiXdhXI0AtA&usqp=CAU',
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRuaVjRvjIZAO6XIsd0eHS22Q949ke7r-KZFg&usqp=CAU',
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRkIVlYNBoYdfq1FGWs_dx_k8Irvjmdu64p7A&usqp=CAU',
        'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRwor6J4qPx1WWN27GYiZEFP-NAoWSstn-YrQ&usqp=CAU',
    ];

    const generateRandomURL = () => {
        const number = Math.random() * 6;
        const index = Math.floor(number);
        return dummyImages[index];
    };
    //theme handling
    const theme = useTheme();
    const styles = useStyles();
    const windowWidth = Dimensions.get('window').width;
    const windowHeight = Dimensions.get('window').height;
    const navigation = useNavigation();
    const isLoading = useSelector((state: ReducerState) => state.loadingReducer.isLoading);
    const favoriteBooks = useSelector((state: ReducerState) => state.appReducer.favorite);
    const { onRefresh } = props;
    const navigateToDetails = async (params: number) => {
        //to check if the internet connection is working
        navigation.navigate('BookDetail' as never, params as never);
    };
    const bookList = (item: FavoriteBook) => {
        return (
            <TouchableHighlight
                key={item.id}
                underlayColor={theme.colors.highlight}
                onPress={() => {
                    navigateToDetails(item.bookId);
                }}>
                <BookCard
                    url={generateRandomURL()}
                    styleSelect="Large"
                    id={item?.book.id}
                    bookTitle={item?.book?.title}
                    book={favoriteBooks}
                    hideIcon={false}
                />
            </TouchableHighlight>
        );
    };
    return (
        <View style={styles.mainView}>
            {favoriteBooks?.length > 0 ? (
                <FlatList
                    numColumns={windowHeight > windowWidth ? 2 : 4}
                    key={windowHeight > windowWidth ? 2 : 4}
                    keyExtractor={(item, index) => 'key' + index}
                    showsVerticalScrollIndicator={false}
                    data={favoriteBooks}
                    style={styles.flatList}
                    contentContainerStyle={{ flexGrow: 1 }}
                    onRefresh={onRefresh}
                    refreshing={isLoading}
                    renderItem={({ item }) => bookList(item)}
                    showsHorizontalScrollIndicator={false}
                />
            ) : (
                <ScrollView contentContainerStyle={styles.favoriteView}>
                    <Image source={images.books.noBookFound} style={styles.imageError} />
                    <Text style={styles.bookmark}>No bookmarks available</Text>
                </ScrollView>
            )}
        </View>
    );
};

export default Container;
