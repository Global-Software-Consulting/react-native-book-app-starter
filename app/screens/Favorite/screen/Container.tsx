//to update the favorites list
import { useNavigation } from '@react-navigation/native';
//importing card component
import BookCard from 'components/BookCard/BookCard';
import images from 'config/images';
import { IBookState } from 'models/reducers/fetchBooks';
import { ILoginState } from 'models/reducers/login';
import React from 'react';
import {
    FlatList,
    Image,
    RefreshControl,
    ScrollView,
    TouchableHighlight,
    View,
} from 'react-native';
import { Text } from 'react-native-paper';
import { heightPercentageToDP } from 'react-native-responsive-screen';
import Toast from 'react-native-simple-toast';
import { useDispatch, useSelector } from 'react-redux';
import NetworkUtils from 'utils/networkUtils';
import { useStyles } from '../styles';
interface Props {
    //     books?: {};
    //     base_url?: string;
    onRefresh: () => void;
}

interface IStateReducer {
    loginReducer: ILoginState;
    appReducer: IBookState;
}

interface IParams {
    id: number;
    bookId: number;
    averageRating: number;
    title: string;
    numberOfPages: string | number;
    shortSummary: string;
    book: {
        title: string;
        id: number;
    };
}

const Container: React.FC<Props> = (props) => {
    //theme handling
    const styles = useStyles();
    const dispatch = useDispatch();
    const navigation = useNavigation();
    const isLoading = useSelector((state: IStateReducer) => state.appReducer.isFetching);
    const favoriteBooks = useSelector((state: IStateReducer) => state.appReducer.favorite);
    const { onRefresh } = props;
    const navigateToDetails = async (params: IParams) => {
        //to check if the internet connection is working
        const isConnected = await NetworkUtils.isNetworkAvailable();
        if (isConnected) {
            navigation.navigate('BookDetail', params);
        } else {
            Toast.show('You are offline', Toast.SHORT);
        }
    };

    const FavoriteBooks = () => {
        return (
            <View>
                <FlatList
                    nestedScrollEnabled={true}
                    numColumns={2}
                    contentContainerStyle={styles.flatList}
                    data={favoriteBooks}
                    renderItem={({ item }) => (
                        <TouchableHighlight
                            key={item}
                            underlayColor="grey"
                            onPress={() => {
                                navigateToDetails(item);
                            }}>
                            <BookCard
                                url={
                                    'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRoE4lMLbADvLAxUvZf5ZAGvHUZ3KpBWFTW1g&usqp=CAU'
                                }
                                styleSelect="Large"
                                id={item?.book.id}
                                bookTitle={item?.book?.title}
                                book={favoriteBooks}
                                hideIcon={false}
                            />
                        </TouchableHighlight>
                    )}
                    showsHorizontalScrollIndicator={false}
                />
            </View>
        );
    };

    return (
        <View style={{ height: heightPercentageToDP('100%') }}>
            <ScrollView
                style={styles.container}
                nestedScrollEnabled
                refreshControl={<RefreshControl refreshing={isLoading} onRefresh={onRefresh} />}>
                {favoriteBooks.length > 0 ? (
                    <FavoriteBooks />
                ) : (
                    <View style={styles.favoriteView}>
                        <Image source={images.books.noBookFound} style={styles.imageError} />
                        <Text style={styles.bookmark}>No bookmarks available</Text>
                    </View>
                )}
            </ScrollView>
        </View>
    );
};

export default Container;
