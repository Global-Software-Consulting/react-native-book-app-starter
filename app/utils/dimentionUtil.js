import { useDeviceOrientation } from '@react-native-community/hooks';
import { useCallback, useEffect, useMemo, useState } from 'react';
import { Dimensions, PixelRatio } from 'react-native';

const widthPercentageToDP = (screenWidth, widthPercent) => {
    const elemWidth = typeof widthPercent === 'number' ? widthPercent : parseFloat(widthPercent);
    return PixelRatio.roundToNearestPixel((screenWidth * elemWidth) / 100);
};

const heightPercentageToDP = (screenHeight, heightPercent) => {
    const elemHeight =
        typeof heightPercent === 'number' ? heightPercent : parseFloat(heightPercent);
    return PixelRatio.roundToNearestPixel((screenHeight * elemHeight) / 100);
};

export function getPercentageWidth() {
    const [width, setWidth] = useState(Dimensions.get('window').width);
    useEffect(() => {
        Dimensions.addEventListener('change', (newDimensions) => {
            // Retrieve and save new dimensions
            screenWidth = newDimensions.window.width;
            setWidth(screenWidth);
        });
        return Dimensions.removeEventListener('change', () => {});
    }, []);
    const wp = useCallback((value) => widthPercentageToDP(width, value), [width]);

    return wp;
}
export function getPercentageHeight() {
    const [height, setHeight] = useState(Dimensions.get('window').height);
    useEffect(() => {
        Dimensions.addEventListener('change', (newDimensions) => {
            screenHeight = newDimensions.window.height;
            setHeight(screenHeight);
        });
        return Dimensions.removeEventListener('change', () => {});
    }, []);
    const hp = useCallback((value) => heightPercentageToDP(height, value), [height]);

    return hp;
}
