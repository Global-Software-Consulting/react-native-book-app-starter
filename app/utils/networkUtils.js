import NetInfo from "@react-native-community/netinfo";
export default class NetworkUtils {
  static async isNetworkAvailable() {
    const response = await NetInfo.fetch();
    return response.isConnected;
}}