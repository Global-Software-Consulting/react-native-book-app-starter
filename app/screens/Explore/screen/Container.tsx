import React, {useEffect, useState} from 'react';
import {useTranslation} from 'react-i18next';
import {
  Alert,
  FlatList,
  Image,
  RefreshControl,
  ScrollView,
  TextInput,
  TouchableHighlight,
  View,
} from 'react-native';
import {Text, useTheme} from 'react-native-paper';
import Icon from 'react-native-vector-icons/MaterialIcons';
import {useDispatch, useSelector} from 'react-redux';
import {useStyles} from 'screens/Explore/styles';
//importing card component
import BookCard from 'components/BookCard/BookCard';
import i18n from 'config/Languages/index';
import * as appActions from 'store/actions/appActions';
import * as loginActions from 'store/actions/loginActions';
import {useNavigation} from '@react-navigation/native';
import AsyncStorage from '@react-native-async-storage/async-storage';
import {
  heightPercentageToDP,
  widthPercentageToDP,
} from 'react-native-responsive-screen';
import images from 'config/images';
import {SafeAreaProvider} from 'react-native-safe-area-context';
import Toast from 'react-native-simple-toast';
import NetworkUtils from 'utils/networkUtils';
import {useIsFocused} from '@react-navigation/native';

interface Props {
  books?: [];
  name?: string;
  base_url?: string;
}
const initI18n = i18n;

const ExploreComponent: React.FC<Props> = ({name, base_url}) => {
  //theme handling
  const styles = useStyles();
  const {t, i18n} = useTranslation();
  const theme = useTheme();
  const dispatch = useDispatch();
  const [searchText, setSearchText] = useState('');
  const navigation = useNavigation();
  //getting data from store
  const books = useSelector(state => state.appReducer.detail);
  const isFocused = useIsFocused();
  const isLoading = useSelector(state => state.appReducer.isFetching);
  const favoriteBooks = useSelector(state => state.appReducer.favorite);
  const token = useSelector(state => state.loginReducer.token);
  const newFavorites: string[] = favoriteBooks;
  const searchBook = (bookName: string) => {
    dispatch(appActions.IFetchBooksRequest(bookName));
  };
  const [username, setUserName] = useState(
    userData.firstName + ' ' + userData.lastName,
  );
  const fetchBookDetails = async () => {
    const isConnected = await NetworkUtils.isNetworkAvailable();
    if (!isConnected) {
      dispatch(appActions.IFetchBooksRequest('a'));
      getFavoriteBooks();
    } else {
      Toast.show('You are offline', Toast.SHORT);
    }
  };

  //fetching favorite books
  const getFavoriteBooks = async () => {
    dispatch(appActions.IFetchFavoriteBooksRequest());
  };

  const navigateToDetails = async params => {
    const isConnected = await NetworkUtils.isNetworkAvailable();
    if (!isConnected) {
      navigation.navigate('BookDetail', params);
    } else {
      Toast.show('You are offline', Toast.SHORT);
    }
  };

  return (
    <View style={styles.mainViewSetting}>
      {books.length > 0 ? (
        <ScrollView
          nestedScrollEnabled={true}
          style={styles.container}
          refreshControl={
            <RefreshControl
              refreshing={isLoading}
              onRefresh={fetchBookDetails}
            />
          }>
          <View>
            {/* Searchbar */}
            <View style={styles.searchView}>
              <TextInput
                underlineColorAndroid="transparent"
                placeholder={t('Search Here')}
                placeholderTextColor={theme.colors.text}
                onChangeText={text => setSearchText(text)}
                style={styles.searchViewChildren}
                onEndEditing={() => searchBook(searchText)}
              />
              <Icon
                name="find-in-page"
                size={30}
                style={styles.searchViewChildren}
                onPress={() => searchBook(searchText)}
              />
            </View>
            <Text style={styles.name}>
              {t('Hi')} {name}{' '}
            </Text>
            <Text style={styles.tagLine}>{t('Lets find something new')}</Text>

            <View style={styles.horizontalRuler} />

            <Text style={styles.listCaption}>{t('Trending')}</Text>
            <FlatList
              nestedScrollEnabled={true}
              horizontal
              data={books?.filter(item => {
                return item?.averageRating > 3;
              })}
              contentContainerStyle={styles.flatList}
              renderItem={({item, index}) => (
                <TouchableHighlight
                  key={item}
                  underlayColor="grey"
                  onPress={() => {
                    navigateToDetails(item);
                  }}>
                  <BookCard
                    url={
                      'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRoE4lMLbADvLAxUvZf5ZAGvHUZ3KpBWFTW1g&usqp=CAU'
                    }
                    isFavorite={
                      newFavorites.indexOf(item.title) !== -1 ? true : false
                    }
                    styleSelect="Custom"
                    bookTitle={item?.title}
                    book={item}
                    id={item?.id}
                  />
                </TouchableHighlight>
              )}
              showsHorizontalScrollIndicator={false}
            />

            <View style={styles.horizontalRuler} />

            <Text style={styles.listCaption}>{t('New Releases')}</Text>
            <FlatList
              horizontal
              data={books?.filter(item => {
                return item?.averageRating <= 3 && item?.averageRating > 0;
              })}
              contentContainerStyle={styles.flatList}
              renderItem={({item, index}) => (
                <TouchableHighlight
                  key={item}
                  underlayColor="grey"
                  onPress={() => {
                    navigateToDetails(item);
                  }}>
                  <BookCard
                    url={
                      'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRoE4lMLbADvLAxUvZf5ZAGvHUZ3KpBWFTW1g&usqp=CAU'
                    }
                    styleSelect="General"
                    bookTitle={item?.title}
                    book={item}
                    id={item?.id}
                  />
                </TouchableHighlight>
              )}
              showsHorizontalScrollIndicator={false}
            />

            <View style={styles.horizontalRuler} />

            <Text style={styles.listCaption}>{t('Selected for you')}</Text>
            <FlatList
              horizontal
              data={books?.filter(item => {
                return item?.averageRating == 0;
              })}
              contentContainerStyle={styles.flatListLast}
              renderItem={({item, index}) => (
                <TouchableHighlight
                  key={item}
                  underlayColor="grey"
                  onPress={() => {
                    navigateToDetails(item);
                  }}>
                  <BookCard
                    url={
                      'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRoE4lMLbADvLAxUvZf5ZAGvHUZ3KpBWFTW1g&usqp=CAU'
                    }
                    isFavorite={
                      newFavorites.indexOf(item?.title) !== -1 ? true : false
                    }
                    styleSelect="General"
                    bookTitle={item?.title}
                    book={item}
                    id={item?.id}
                  />
                </TouchableHighlight>
              )}
              showsHorizontalScrollIndicator={false}
            />
          </View>
        </ScrollView>
      ) : (
        <View style={styles.favoriteView}>
          <Image source={images.books.noBookFound} style={styles.imageError} />
          <Text style={styles.bookmark}>
            No books available or check your internet
          </Text>
        </View>
      )}
    </View>
  );
};

export default ExploreComponent;
