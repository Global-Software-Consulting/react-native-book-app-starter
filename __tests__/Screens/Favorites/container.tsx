import { useIsFocused, useNavigation } from '@react-navigation/native'
import { fireEvent, render } from '@testing-library/react-native'
import React from 'react'
import * as redux from 'react-redux'

import { act } from 'react-test-renderer'
import Container from 'screens/Favorite/Container'
import * as utils from 'utils/dimentionUtil'
import { getPercentageHeight, getPercentageWidth } from 'utils/dimentionUtil'
const DeviceTypeUtilsMock = jest.requireMock('utils/dimentionUtil')
import 'react-native-gesture-handler/jestSetup'

jest.mock('react-native-reanimated', () => {
    const Reanimated = require('react-native-reanimated/mock')

    Reanimated.default.call = () => {}

    return Reanimated
})

// Silence the warning: Animated: `useNativeDriver` is not supported because the native animated module is missing
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper')

jest.mock('./../../../app/utils/dimentionUtil', () => {
    return {
        getPercentageHeight: jest.fn(),
        height: '100',
        getPercentageWidth: jest.fn(),
    }
})

jest.mock('@react-navigation/core', () => {
    const actualNav = jest.requireActual('@react-navigation/core')
    return {
        ...actualNav,
        useNavigation: () => ({
            navigate: jest.fn(),
            dispatch: jest.fn(),
        }),
        useIsFocused: jest.fn(),
    }
})
jest.mock('react-native-image-crop-picker', () => {
    return {
        __esModule: true,
        A: true,
        useDispatch: jest.fn(),
        default: 'mockedDefaultExport',
    }
})

jest.mock('react-native-keyboard-aware-scroll-view', () => {
    return {
        __esModule: true,
        A: true,
        KeyboardAwareScrollView: jest.fn(),
        default: 'mockedDefaultExport',
    }
})
jest.mock('react-native-linear-gradient', () => {
    return {
        __esModule: true,
        A: true,
        LinearGradient: jest.fn(),
        default: 'mockedDefaultExport',
    }
})
jest.mock('./../../../app/config/images', () => {
    return {
        __esModule: true,
        A: true,
        app: { logo: jest.fn() },
        default: 'mockedDefaultExport',
    }
})

jest.mock('react-native-tts', () => {
    return {
        __esModule: true,
        A: true,
        Tts: jest.fn(),
        default: 'mockedDefaultExport',
    }
})

describe('Screen test', () => {
    test('Snapshot ', () => {
        const height = jest.fn()
        getPercentageHeight.mockReturnValue(jest.fn())
        const width = jest.fn()
        getPercentageWidth.mockReturnValue(jest.fn())

        const isFocused = jest.fn()
        useIsFocused.mockReturnValue(true)

        act(() => {
            const state = {
                loadingReducer: { isLoading: false },
                appReducer: { favorite: 'as' },
            }
            jest.spyOn(redux, 'useSelector').mockImplementation((val) => {
                val(state)
                expect(tree).toMatchSnapshot()
            })
        })
        const tree = render(<Container />)
    })

    test('Snapshot ', () => {
        const height = jest.fn()
        getPercentageHeight.mockReturnValue(jest.fn())
        const width = jest.fn()
        getPercentageWidth.mockReturnValue(jest.fn())
        act(() => {
            const state = {
                loadingReducer: { isLoading: false },
                appReducer: { favorite: ['as', 'bd', 'ef'] },
            }
            jest.spyOn(redux, 'useSelector').mockImplementation((val) => {
                val(state)
            })
        })
        const isFocused = jest.fn()
        useIsFocused.mockReturnValue(true)

        const { getByTestId } = render(<Container toDos={true} />)
    })
})
