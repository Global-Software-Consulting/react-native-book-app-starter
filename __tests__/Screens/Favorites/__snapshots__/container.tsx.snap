// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen test Snapshot  1`] = `undefined`;

exports[`Screen test Snapshot  2`] = `undefined`;
